actors/migration/top.go

```diff
@@ -249,11 +328,32 @@ func migrateOneActor(ctx context.Context, store cbor.IpldStore, addr address.Add
 		CallSeqNum: actorIn.CallSeqNum,
 		Balance:    big.Add(actorIn.Balance, transfer),
 	}
-	err = actorsOut.SetActor(addr, &actorOut)
-	if err != nil {
-		return big.Zero(), err
+	syncCtx.actorsOutCh <- actorUpdate{addr: addr, actorOut: actorOut}
+	if transfer.GreaterThan(big.Zero()) {
+		syncCtx.transferCh <- transfer
+	}
+}
+
+func accumulateResults(syncCtx *SyncContext, transferAcc big.Int, powerUpdatesAcc *PowerUpdates, actorsOut *states.Tree) error {
+	// Read and process worker results from all channels
+READLOOP:
+	for {
+		select {
+		case err := <-syncCtx.errCh:
+			return err
+		case transfer := <-syncCtx.transferCh:
+			transferAcc = big.Add(transferAcc, transfer)
+		case pwrUpdate := <-syncCtx.powerUpdateCh:
+			pwrUpdate.Apply(powerUpdatesAcc)
+		case actorUpdate := <-syncCtx.actorsOutCh:
+			if err := actorsOut.SetActor(actorUpdate.addr, &actorUpdate.actorOut); err != nil {
+				return err
+			}
+		default:
+			break READLOOP
```

Comment:
It would be nice to accumulate results in a different goroutine, rather than start/stop like this. I think this works, but it's hard to reason about.

Reply:
I hear you.  One advantage of the current approach is the simplicity of error handling in the main goroutine.  I played with reading the error channel in the main loop and sending directly there from the worker goroutines which is nice.  But an annoying thing is that the processing goroutine can also error, so I can't think of a better way forward than having two error channels and routing all errors through the processing goroutine which is starting to make this change feel not worth it.  Am I missing any clean patterns for dealing with this?

Reply:
The standard pattern here is to create a fan-out/fan-in pipeline.

1. Have one goroutine generating work, sending it to an "input" channel.
2. Have N (workers) goroutines processing work, sending results to an "output" channel.
3. Back on the main goroutine, process results off the "output" channel.

There are a couple of ways to do this, but the errgroup library makes this pretty simple:



```go
package main

import (
	"context"
	"sync"

	"golang.org/x/sync/errgroup"
)

const workers = 10

type job struct{}
type result struct{}

func getInput(ctx context.Context) (job, error)           { panic("unimplemented") }
func doWork(ctx context.Context, inp job) (result, error) { panic("unimplemented") }
func applyResult(ctx context.Context, res result) error   { panic("unimplemented") }

func process(ctx context.Context) (int, error) {
	grp, ctx := errgroup.WithContext(ctx)

	input := make(chan job)
	output := make(chan result)

	// Pipeline stage 1: generate jobs (fan out)

	grp.Go(func() error {
		// signal to the workers that there is no new input.
		defer close(input)

		for {
			nextInput, err := getInput(ctx)
			if err != nil {
				return err
			}

			select {
			case input <- nextInput:
			case <-ctx.Done():
				return ctx.Err()
			}
		}
	})

	// Pipeline stage 2: process jobs (map).

	var workerWg sync.WaitGroup
	for i := 0; i < workers; i++ {
		workerWg.Add(1)
		grp.Go(func() error {
			defer workerWg.Done()

			for inp := range input {
				res, err := doWork(ctx, inp)
				if err != nil {
					return err
				}
				select {
				case output <- res:
				case <-ctx.Done():
					return ctx.Err()
				}
			}
			return nil
		})
	}
	// standard pattern for closing an output channel once workers are done.
	grp.Go(func() error {
		workerWg.Wait()
		close(output)
		return nil
	})

	// Pipeline stage 3: apply results (fan in).

	// This doesn't _have_ to go in a separate goroutine, but it can
	// simplify error handling.
	var res int
	grp.Go(func() error {
		for res := range output {
			if err := applyResult(ctx, res); err != nil {
				return err
			}
		}
		return nil
	})

	if err := grp.Wait(); err != nil {
		return 0, err
	}

	return res, nil
}
```

Reply:
Thank you for the detailed explanation